src/object/cli_obj.c

```diff
@@ -896,6 +896,8 @@ struct obj_auxi_args {
 	tse_task_t			*obj_task;
 	struct daos_obj_shard_tgt	*fw_shard_tgts;
 	uint32_t			 fw_cnt;
+	uint32_t			 bulk_nr;
+	crt_bulk_t			*bulks;
```

Comment:
It looks to me that bulk handle is per RPC stuff, so probably we'd keep it in shard layer?

With EC introduced, single akey fetch/update split into fetch/update over multiple shards, and each shard will have it's own different bulk handle (not quite sure if my assumption on this EC io split is correct)

I have no expertise on client io nor cart, so please correct me if I misunderstand anything important.

Reply:
I think we'd better share the same bulk handle, because memory region registration is quite expensive, it's better to share the same bulk handle if possible.

Reply:
for memory registration overhead, it is very related with provider.
Now for sockets and psm2 provider, the mem reg overhead basically can be ignored because it actually does nothing. The page pin is dynamically done on-the-fly when bulk transferring.
For verbs provider, I heard that mem reg with quite big overhead. I guess it related with the IB verbs/driver need to walk through page table and pin each page.

But creating one bulk handle for one IOD at obj layer, and let different targets choose applicable portion of the source bulk handle seems reasonable to me, despite of which provider uses.